pallets/gear/src/tests.rs

```diff
@@ -6982,6 +6982,174 @@ fn system_reservation_zero_amount_panics() {
     });
 }
 
+#[test]
+fn gas_reservation_works() {
+    use demo_reserve_gas::{HandleAction, InitAction};
+
+    init_logger();
+    new_test_ext().execute_with(|| {
+        assert_ok!(Gear::upload_program(
+            RuntimeOrigin::signed(USER_1),
+            demo_reserve_gas::WASM_BINARY.to_vec(),
+            DEFAULT_SALT.to_vec(),
+            InitAction::Normal.encode(),
+            10_000_000_000,
+            0,
+        ));
+
+        let pid = get_last_program_id();
+
+        run_to_block(2, None);
+
+        // gas has been reserved 3 times
+        let map = get_reservation_map(pid).unwrap();
+        assert_eq!(map.len(), 3);
+
+        let GasInfo {
+            min_limit: spent_gas,
+            ..
+        } = Gear::calculate_gas_info(
+            USER_1.into_origin(),
+            HandleKind::Handle(pid),
+            HandleAction::Unreserve.encode(),
+            0,
+            true,
+        )
+        .expect("calculate_gas_info failed");
+
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_1),
+            pid,
+            HandleAction::Unreserve.encode(),
+            spent_gas,
+            0
+        ));
+
+        run_to_block(3, None);
+
+        // gas unreserved manually
+        let map = get_reservation_map(pid).unwrap();
+        assert_eq!(map.len(), 2);
+
+        run_to_block(2 + 2, None);
+
+        // gas not yet unreserved automatically
+        let map = get_reservation_map(pid).unwrap();
+        assert_eq!(map.len(), 2);
+
+        run_to_block(2 + 3, None);
+
+        // gas unreserved automatically
+        let map = get_reservation_map(pid).unwrap();
+        assert_eq!(map.len(), 1);
+
+        // check task is exist yet
+        let (reservation_id, slot) = map.iter().next().unwrap();
+        let task = ScheduledTask::RemoveGasReservation(pid, *reservation_id);
+        assert!(TaskPoolOf::<Test>::contains(
+            &BlockNumberFor::<Test>::from(slot.finish),
+            &task
+        ));
+
+        // `gr_exit` occurs
```

Comment:
just for confidence, do we have test for wait+wake in init and further termination for common reservations and system one?

Reply:
and gr_exit test for system reservation with a)wait+wake and b)during the single execution